### PR TITLE
feature/add-featured-image-management

### LINKED ISSUE #169: Can't remove a featured image from a post
Currently you can't remove a featured image from a post, only add. Perhaps we should make it more like WPiOS where there is a separate section in the post editor for adding/updating/removing a featured image.

### LINKED ISSUE #2456: Featured images not working when selected from WP Media Library
1. Create a new post with the app.
2. Tap the image icon and choose “Select from media library” or choose an image from the "Site images" tab with the media picker (for 3.9+)
3. Choose an image then tap it to edit and tick the box next to “Use as featured image”.
4. Publish the post.

### PR DESCRIPTION
**Pull request by fadils**
Fix #169 
First pass.

Not as exactly as initial plan to have a central featured image management in Settings.
For this, probably need some help in changing adding a featured_image column in posts table.
I did it manually, resulting in db versioning mess up.
I also think this would be appropriate in a separate PR.

Once we have a featured_image column, we probably can fix #2456 too.

Will squash if everything is OK.

### CODE REVIEW

**Comment by maxme** [comment]
IMO, the UX is not good enough.

The previous UX (before your patch) was bad because:

1. Most users don't know they can set a featured image by tapping the edit button on included pictures. We can fix it by having "Select featured image button" in the Settings screen.
1. When they know they can select featured images, they don't know if the image will also be included in the post. We can fix this by having a strong distinction between included images and featured image. (ie remove the "Use as featured image" button in the image edit screen).

**Comment by maxme** [comment]
> For this, probably need some help in changing adding a featured_image column in posts table.
> I did it manually, resulting in db versioning mess up.

You can add DB migration to the `posts` table in `WordPressDB`, like we did here: https://github.com/wordpress-mobile/WordPress-Android/blob/08765f5408bf3aaceb7a0a4b01f823eb8c87ccdc/WordPress/src/main/java/org/wordpress/android/WordPressDB.java#L279-L282

**Comment by fadils** [comment]
@maxme thanks!

1. I agree, having it centralized is the way to go..
So, should we merge this first and create another PR for centralized featured image setting?

2. Will do.

**Comment by maxme** [comment]
> So, should we merge this first and create another PR for centralized featured image setting?

No, we're going to merge it when it's finished.

**Comment by maxme** [comment]
UX looks good. Would be great to:

1. Remove video tabs from the media picker.
1. Add a new tab: "Post images" listing images already included in the post.

There is a major issue: that won't work on self hosted sites since you're using the REST API only (self hosted sites use XMLRPC).

There a few checks to `isAdded()` missing in `EditPostSettingsFragment` that can make the app crash if the fragment isn't attached.

Local drafts doesn't save the selected featured image: Create a new post, set a featured image, go back to the post list, re open that post, you won't see the selected FI.

**Comment by fadils** [comment]
> UX looks good.

Thanks for your direction. :grin: 

> Remove video tabs from the media picker

Will do. Already removed the "Video" button on the actionBar. Forgot to remove videos on the tabs.
Thanks for the reminder.

> Add a new tab: "Post images" listing images already included in the post.

Will do. Awesome idea. :clap: 

> There is a major issue: that won't work on self hosted sites since you're using the REST API only (self hosted sites use XMLRPC).

I'll message you on Slack about this when you're online.

> There a few checks to isAdded() missing in EditPostSettingsFragment that can make the app crash if the fragment isn't attached.

https://github.com/fadils/WordPress-Android/blob/issue/169-add-featured-image-removal/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java#L134
I already instantiated activity in `onAttach`, and use that instance instead of `getActivity()` to prevent `NullPointerException`. Do I still need to call `isAdded()`?

> Local drafts doesn't save the selected featured image: Create a new post, set a featured image, go back to the post list, re open that post, you won't see the selected FI.

Probably the featured image is not set yet.. (still uploading if you happen to choose from local image). Then yes it won't be shown (or even associated). I already put a loading indicator on the center of featured image and put a toast after it's successfully set. Looking forward to another of your input, though. :grin:

**Comment by maxme** [comment]
> Do I still need to call isAdded()?

For example, the app can crash here: https://github.com/fadils/WordPress-Android/blob/issue/169-add-featured-image-removal/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java#L1180

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java [hunk 1]
```diff
@@ -875,4 +962,286 @@ private void populateSelectedCategories() {
...
+
+    /**
+     * Create image {@link org.wordpress.mediapicker.source.MediaSource}'s for media selection.
+     *
+     * @return
+     *  list containing all sources to gather image media from
+     */
+    private ArrayList<MediaSource> imageMediaSelectionSources() {
+        ArrayList<MediaSource> imageMediaSources = new ArrayList<>();
+        imageMediaSources.add(new MediaSourceDeviceImages(getActivity().getContentResolver()));
```
Comment (tonyr59h) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java]: Just a heads up that the MediaPicker version on develop has been bumped to v1.2.1 and this interface has changed. Ping me on Slack if you run into any troubles with that.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java [hunk 2]
```diff
@@ -875,4 +962,286 @@ private void populateSelectedCategories() {
...
+        }
+        mActivity.mSectionsPagerAdapter.notifyDataSetChanged();
+    }
+
+    @Override
+    public void updateFragment() {
+        mSetFeaturedImageButton.setVisibility(View.GONE);
+        mRemoveFeaturedImageButton.setVisibility(View.VISIBLE);
+
+        if (mFeaturedImageURL.contains("wordpress.com")) {
```
Comment (tonyr59h) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java]: This results in a crash when selecting media from a self-hosted site. Take a look at #2509 for reference.

Comment (fadils) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java]: Thanks!
:grin:

**Comment by maxme** [comment]
* Featured images in local drafts still don't work.
* Create a post on the Web with a FI, Open it in the app: featured image is not there.

**Comment by maxme** [comment]
I think you should rethink the update flow. Do you need to do a network call when setting a new featured image? Can't you store the state locally and only make the change when the user push the "Publish" (or "Update") button in the editor?

**Comment by tonyr59h** [comment]
The Gallery menu item is still available when in mutl-select mode. Clicking it doesn't do anything though, would be good to remove it altogether ([here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/media/MediaPickerActivity.java#L511)).

**Comment by tonyr59h** [comment]
On second thought we should prevent the user from entering multi-select mode altogether. If you disable the animations in onActionModeStarted and onActionModeFinished (but save mActionMode) you can immediately exit it in [onMediaSelected](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/media/MediaPickerActivity.java#L231).

**Comment by fadils** [comment]
Thanks @tonyr59h for the pointer.

**Comment by fadils** [comment]
Post images tab is added. One currently known limitation is to make the Editor aware if a user deletes a media.

Need help from @maxme and @tonyr59h to review it. Would love to hear further inputs from you guys. :grin: 

cc: @beaucollins

**Comment by maxme** [comment]
Note: don't merge this before the new media picker is the only solution to select images in a post (we re introduced the old ways in 4.0).

**Pull request closed without merging**